llvm/lib/Target/RISCV/MCTargetDesc/RISCVAsmBackend.cpp

```diff
@@ -660,6 +699,14 @@ bool RISCVAsmBackend::addReloc(const MCFragment &F, const MCFixup &Fixup,
   if (IsResolved &&
       (getFixupKindInfo(Fixup.getKind()).Flags & MCFixupKindInfo::FKF_IsPCRel))
     IsResolved = isPCRelFixupResolved(Target.getAddSym(), F);
+
+  // Some Fixups require a vendor relocation, record it (directly) before we add
+  // the relocation.
+  if (!IsResolved || shouldForceRelocation(Fixup, Target)) {
```

Comment:
If we are going to call `shouldForceRelocation`, just avoid the generic `MCAsmBackend::addReloc` and call the underlying `recordRelocation` directly.